subprojects/docs/src/snippets/dependencyManagement/customizingResolution-attributeSubstitutionRule/kotlin/consumer/build.gradle.kts

```diff
@@ -30,8 +30,9 @@ configurations.all {
 // end::substitution_rule_alternative[]
 
 tasks.register("resolve") {
-    inputs.files(configurations.runtimeClasspath)
+    val classpath: Provider<out FileCollection> = configurations.runtimeClasspath
```

Comment:
❓ Is `out` required here?

Reply:
> Is `out` required here?

Yes, as the RHS is of type `NamedDomainObjectProvider<Configuration>`. It is a subtype of `Provider<Configuration>`, but `Provider` is a Java generic type, therefore it is invariant, so the value cannot be directly assigned to plain `Provider<FileCollection>`. We need either an unchecked cast, a `.map { it }` or use-site variance declaration, I opted for the latter.

Reply:
I see, thanks, @mlopatkin. 👍

Reply:
That would make it type `Provider<Configuration>` which would be incompatible with configuration cache afair.
I think this should work:
```
val runtimeClasspath: Provider<out FileCollection> by configurations.existing
```
but then you win nothing as there is a type-safe accessor that does it for you.

Reply:
Here's a video answer by @bamboo : https://youtu.be/3L7-pGdCK3E?t=4101
Unfortunately, it does not work even with `val runtimeClasspath: Provider<out FileCollection> by configurations.existing`

Reply:
Thanks for bringing that up during the event, @vlsi. The Kotlin compiler is generating a much more complex structure for it than I naively assumed.

Reply:
However, then I wonder what are the benefits of doing `.get()` within `doLast`.
Why don't we use `val runtimeClasspath: FileCollection = configurations.runtimeClasspath.get()` here?
See https://github.com/gradle/gradle/pull/21829/files#r960950954

Then both Groovy and Kotlin would use pretty much the same APIs.
Currently, the PR uses non-lazy for Groovy and lazy for Kotlin which is misleading, especially for those who use "Groovy-Kotlin" switcher to learn Gradle DSL

Reply:
Here's yet another pattern (`configurations.getByName(String)`): https://github.com/gradle/gradle/blob/f632e1ac374903c849de79246bc7b904f45df013/subprojects/docs/src/snippets/testing/test-suite-multi-configure-each/kotlin/build.gradle.kts#L51-L53, and I think the DSL should favour `configurations["integrationTestRuntimeClasspath"]` instead of `configurations.getByName("integrationTestRuntimeClasspath")` in the case when accessors are not available.

Reply:
I think, as Vladimir suggested, that the following is the simplest:
```
val runtimeClasspath: FileCollection = configurations.runtimeClasspath.get()
```
Configurations are created eagerly anyway so using the provider doesn't make much sense. Moreover, `FileCollection` is already a lazy type so there's no need to wrap it in a provider.

Reply:
Suggestion accepted in https://github.com/gradle/gradle/pull/21866.